retrieval/discovery/kubernetes/node.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/util/strutil"
+	"golang.org/x/net/context"
+)
+
+type nodeDiscovery struct {
+	mtx           sync.RWMutex
+	nodes         map[string]*Node
+	retryInterval time.Duration
+	kd            *Discovery
+}
+
+func (d *nodeDiscovery) run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	nodes, _, err := d.getNodes()
```

Comment:
As long as (speaking in generic terms) watchARTIFACT() definitely emits the proper events when initializing the watch, i.e., _without_ receive an updated ARTIFACT notification, then doing it also at the beginning of run() would indeed be duplicative.

p.s., my kingdom for generics.